File: tracker_parser.cpp
#include "tracker_parser.h"

int TrackerFile::hostTTL = 15; //Default 15 min until delete hosts

TrackerFile::TrackerFile()
{
  m_filename    = "";
  m_filesize    = 0;
  m_description = "";
  m_md5         = "";
}

string TrackerFile::getFilename()
{
  return m_filename;
}

long TrackerFile::getFilesize()
{
  return m_filesize;
}

string TrackerFile::getDescription()
{
  string retVal = m_description;

  for(int i=0,size=retVal.size(); i<size; i++)
    if(retVal[i] == ' ')
      retVal[i] = '_';

  if(retVal.size()==0)
    retVal = '_'; 
 
  return retVal;
}

string TrackerFile::getMD5()
{
  return m_md5;
}

bool TrackerFile::parseTrackerFile(const char* trackerFileName)
{
  //Read file into fileContents
  long   size         = 0;
  size_t length       = 0;
  char*  fileContents = NULL;
  FILE*  trkrFile     = fopen(trackerFileName,"r");
  bool   error        = false;

  if(trkrFile!=NULL)
  {
    if((fseek(trkrFile, 0L, SEEK_END) == 0) &&
      ((size = ftell(trkrFile)) != -1))
    {
      fileContents = new char[size+1];
      if(fseek(trkrFile, 0L, SEEK_SET) == 0)
      {
        length = fread(fileContents, sizeof(char), size, trkrFile);

        if(length != 0) fileContents[size] = '\0';
        else error = true;
      }
      else error = true;
    }
    else error = true;
    fclose(trkrFile);
  }
  else error = true;

  //Parse File
  stringstream convert;
  char* pch  = NULL;
  char* line = fileContents;

  if(error == 0)
  {
    find<string>("Filename",fileContents,m_filename);
    find<long>("Filesize",fileContents,m_filesize);
    find<string>("Description",fileContents,m_description);
    find<string>("MD5",fileContents,m_md5);

    pch = strtok(fileContents,"\n\0");
    while(pch != NULL)
    {
      parseHost(pch);
      line += strlen(pch)+1;
      pch = strtok(line,"\n\0");
    }
  }
  else
  {
    printf("TrackerFile::parseTrackerFile   There was an error reading file\n");
  }

  if(fileContents!=NULL) delete [] fileContents;
  return error;
}

bool TrackerFile::parseHost(char* line)
{
  vector<string> host;
  char* pch = NULL;

  //Parse 
  pch = strtok(line,":\n\0");
  while(pch != NULL)
  {
    if(strlen(pch) > 0)
      host.push_back(string(pch));
    pch = strtok(NULL,":\n\0");
  }

  //Add host to global
  if(host.size() == 5)
  {
    HostInfo hi;
    stringstream convert;

    for(int i=0,size=host.size(); i<size; i++)
      convert << host[i] << " ";

    convert >> hi.ipaddr
            >> hi.port
            >> hi.startByte
            >> hi.endByte
            >> hi.timestamp;

    m_hosts.push_back(hi);
    return true;
  }

  return false;
}

int TrackerFile::update(const char* cmd)
{
  //If no command then remove old hosts;
  if(cmd == NULL)
  {
    rewriteFile();
    return FILE_SUCC;
  }

  stringstream ss(cmd);
  string trackFile;
  FILE* existingFile;

  long startByte, endByte, timestamp;
  string ipaddr;
  int port;

  int index = -1;

  //Get rid of updateTracker command and copy filename
  ss >> trackFile >> trackFile;
  trackFile += ".track";

  //Check if file exists
  if(existingFile = fopen(trackFile.c_str(), "r"))
  {
    fclose(existingFile);

    //Parse Command
    ss >> startByte >> endByte >> ipaddr >> port;
    timestamp = time(NULL);

    parseTrackerFile(trackFile.c_str());

    //Modify existing tracker information
    for(int i=0,size=m_hosts.size(); i<size; i++)
      if(m_hosts[i].ipaddr == ipaddr)// && m_hosts[i].port == port)
        index = i;

    if(index != -1)
    {
      m_hosts[index].startByte = startByte;
      m_hosts[index].endByte   = endByte;
      m_hosts[index].timestamp = timestamp;
    }
    else
    {
      HostInfo newHost;
      newHost.ipaddr    = ipaddr;
      newHost.port      = port;
      newHost.startByte = startByte;
      newHost.endByte   = endByte;
      newHost.timestamp = timestamp;
      m_hosts.push_back(newHost);
    }

    //Remove out-dated host info and rewrite file
    rewriteFile();

    return FILE_SUCC;  
  }
  return FILE_ERR;
}

int TrackerFile::create(const char* cmd)
{
  stringstream ss(cmd);
  string trackFile;
  FILE* existingFile;

  long startByte, endByte, timestamp;
  string ipaddr;
  int port;

  int index = -1;

  //Get rid of updateTracker command and copy filename
  ss >> m_filename >> m_filename;
  trackFile = m_filename + ".track";

  //Check if file exists
  if(existingFile = fopen(trackFile.c_str(), "r"))
  {
    fclose(existingFile);
    return FILE_ERR;
  }

  //Parse Command
  ss >> m_filesize >> m_description >> m_md5 >> ipaddr >> port;

  startByte = 0;
  endByte   = m_filesize;
  timestamp = time(NULL);

  for(int i=0,size=m_description.size(); i<size; i++)
    if(m_description[i] == '_')
      m_description[i] = ' ';
  
  HostInfo hi;
  hi.ipaddr    = ipaddr;
  hi.port      = port;
  hi.startByte = startByte;
  hi.endByte   = endByte;
  hi.timestamp = timestamp;
  m_hosts.push_back(hi);

  //Remove out-dated host info and rewrite file
  rewriteFile();

  return FILE_SUCC;
}

const HostInfo& TrackerFile::operator[](int i) const
{
  return m_hosts[i];
}

int TrackerFile::getNumHosts() const
{
  return m_hosts.size();
}

void TrackerFile::rewriteFile()
{
  if(m_filename.size() > 0)
  {
    //Remove outdated hosts
    removeHosts();

    //Rewrite tracker file
    string trackFile = m_filename + ".track";
    ofstream outFile(trackFile.c_str());

    outFile << "Filename: "      << m_filename
            << "\nFilesize: "    << m_filesize
            << "\nDescription: " << m_description
            << "\nMD5: "         << m_md5 << endl;
    for(int i=0,size=m_hosts.size(); i<size; i++)
    {
      outFile << m_hosts[i].ipaddr    << ":"
              << m_hosts[i].port      << ":"
              << m_hosts[i].startByte << ":"
              << m_hosts[i].endByte   << ":"
              << m_hosts[i].timestamp << endl;
    }

    outFile.close();
  }
}

void TrackerFile::removeHosts()
{
  long currentTime = time(NULL);
  vector<HostInfo>::iterator it;
  double dt;

  //Remove hosts is timestamp older than hostTTL minutes
  for(it=m_hosts.begin(); it!=m_hosts.end();)
  {
    if((hostTTL*60.0) <= (dt = difftime(currentTime,(*it).timestamp)))
    {
      it = m_hosts.erase(it);
    }
    else
    {
      it++;
    }
  }
}

File: tracker_parser.py
import os
import re
import time

FILE_SUCC = 0
FILE_FAIL = 1
FILE_ERR = 2


class HostInfo():
    def __init__(self):
        self.ip_addr = ''
        self.port = None
        self.start_byte = None
        self.end_byte = None
        self.time_stamp = None


class TrackerFile():
    def __init__(self):
        self.host_TTL = 15
        self.tracker_directory = ''
        self.file_directory = ''
        self.filename = ''
        self.file_size = 0
        self.description = ''
        self.md5 = ''
        self.hosts = []

    def parseTrackerFile(self, tracker_filename):
        error = False

        if not os.path.isfile(tracker_filename):
            error = True
            print "Tracker file does not exist"
            return error

        tracker_file = open(tracker_filename, 'r')
        file_contents = tracker_file.read()
        tracker_file.close()

        match = re.match(r"Filename:\s([^\n]+)", file_contents)
        if match:
            self.filename = match.group(1)
        match = re.match(r"Filesize:\s([^\n]+)", file_contents)
        if match:
            self.file_size = long(match.group(1))
        match = re.match(r"Description:\s([^\n]*)", file_contents)
        if match:
            self.description = match.group(1)
        match = re.match(r"MD5:\s([^\n]+)", file_contents)
        if match:
            self.md5 = match.group(1)

        for line in file_contents.split('\n'):
            host = re.match(r"^([^:\s]+):([^:\s]+):([^:\s]+):([^:\s]+):([^:\s]+)", line)
            if host:
                new_host = HostInfo()
                new_host.ip_addr = host.group(1)
                new_host.port = int(host.group(2))
                new_host.start_byte = long(host.group(3))
                new_host.end_byte = long(host.group(4))
                new_host.time_stamp = long(host.group(5))
                self.hosts.append(new_host)

        if error:
            print "There was an error reading file"
        return error


    def update(self, cmd=''):
        # If no command, remove old hosts
        if cmd == '':
            self._rewrite_file()
            return FILE_SUCC

        tokens = cmd.split(' ')
        if len(tokens) != 6:
            print "Improper update command given"
            return FILE_ERR
        tracker_filename = tokens[1]

        if tracker_filename.find('.track') == -1:
            tracker_filename += '.track'

        if not os.path.isfile(tracker_filename):
            print "Tracker file does not exist"
            return FILE_ERR


        start_byte = tokens[2]
        end_byte = tokens[3]
        ip_address = tokens[4]
        port = tokens[5]
        timestamp = long(time.time())  # whole number of seconds since epoch

        self.parseTrackerFile(tracker_filename)

        update_hosts = [host for host in self.hosts if host.ip_address == ip_address]
        if len(update_hosts) > 0:
            update_hosts[0].start_byte = start_byte
            update_hosts[0].end_byte = end_byte
            update_hosts[0].time_stamp = timestamp
        else:
            new_host = HostInfo()
            new_host.ip_addr = ip_address
            new_host.port = port
            new_host.start_byte = start_byte
            new_host.end_byte = end_byte
            new_host.time_stamp = timestamp
            self.hosts.append(new_host)

        self._rewrite_file()
        return FILE_SUCC

    def create(self, cmd):
        pass

    def updateCommand(self, filename, port):
        pass

    def createCommand(self, filename, port, description="_"):
        pass

    def get_filename(self):
        return self.filename

    def get_file_size(self):
        return self.file_size

    def get_description(self):
        if self.description == '':
            self.description = '_'
        self.description.replace(' ', '_')
        return self.description

    def get_md5(self):
        return self.md5

    def __getitem__(self, item):
        return self.hosts[item]

    def get_num_hosts(self):
        return len(self.hosts)

    def _find(self, key, file_contents, value):
        pass

    def _parse_host(self, line):
        pass

    def _rewrite_file(self):
        pass

    def _remove_hosts(self):
        pass



